### PR TITLE
Allow changing per_connection_buffer_limit_bytes using contour configuration for listeners

### LINKED ISSUE #1407: Lower the per connection buffer to 64k at the listener level
Envoy assigns a 1Mb buffer to to each incoming connection. It is not clear if it consumes the whole 1Mb in one go, or that is a upper limit.

Either way, I think this should be lowered to something like 64. 

https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/lds.proto#envoy-api-field-listener-per-connection-buffer-limit-bytes

### PR DESCRIPTION
**Pull request by rajatvig**
Fixes https://github.com/projectcontour/contour/issues/5263
Fixes #1407

### CODE REVIEW

File: internal/xdscache/v3/listener.go [hunk 1]
```diff
@@ -383,6 +387,7 @@ func (c *ListenerCache) OnChange(root *dag.DAG) {
 				ServerHeaderTransformation(cfg.ServerHeaderTransformation).
 				NumTrustedHops(cfg.XffNumTrustedHops).
 				MaxRequestsPerConnection(cfg.MaxRequestsPerConnection).
+				PerConnectionBufferLimitBytes(cfg.PerConnectionBufferLimitBytes).
```
Comment (skriss) [inline internal/xdscache/v3/listener.go:390]: There are two more places in this file where an `HTTPConnectionManagerBuilder` is used and needs this to be called (one for TLS, one for TLS+fallback)

**Review by skriss — CHANGES_REQUESTED** [review]

File: internal/xdscache/v3/listener.go (hunk 1 above)
Comment (rajatvig) [reply internal/xdscache/v3/listener.go:390]: Updated

File: internal/envoy/v3/listener.go [hunk 2]
```diff
@@ -279,6 +280,12 @@ func (b *httpConnectionManagerBuilder) MaxRequestsPerConnection(maxRequestsPerCo
 	return b
 }
 
+// PerConnectionBufferLimitBytes the soft limit on size of the listener’s new connection read and write buffers
+func (b *httpConnectionManagerBuilder) PerConnectionBufferLimitBytes(perConnectionBufferLimitBytes *uint32) *httpConnectionManagerBuilder {
+	b.perConnectionBufferLimitBytes = perConnectionBufferLimitBytes
```
Comment (skriss) [inline internal/envoy/v3/listener.go:285]: The `Get` method also needs to be modified to generate the xds config for this setting

File: site/content/docs/main/configuration.md [hunk 3]
```diff
@@ -188,9 +188,13 @@ The network configuration block can be used to configure various parameters netw
...
+|-----------------------------------|--------|---------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| connection-balancer               | string | `""`    | This field specifies the listener connection balancer. If the value is `exact`, the listener will use the exact connection balancer to balance connections between threads in a single Envoy process. See [the Envoy documentation][14] for more information. |
+| max-requests-per-connection       | int    | none    | This field specifies the maximum requests for upstream connections. If not specified, there is no limit                                                                                                                                                       |
```
Comment (sunjayBhatia) [inline site/content/docs/main/configuration.md:194]:
```suggestion
| max-requests-per-connection       | int    | none    | This field specifies the maximum requests for downstream connections. If not specified, there is no limit                                                                                                                                                       |
```

File: internal/envoy/v3/listener.go (hunk 2 above)
Comment (skriss) [reply internal/envoy/v3/listener.go:285]: Yep, unfortunately we've got a lot of this

Comment (rajatvig) [reply internal/envoy/v3/listener.go:285]: Had been thinking why nothing failed and suspected I am missing something. Will take a look and change.

Comment (rajatvig) [reply internal/envoy/v3/listener.go]: Changed the code to set it on the Listener

**Review by skriss — APPROVED** [review]
LGTM, thanks @rajatvig!

**Review by sunjayBhatia — APPROVED** [review]

**Pull request merged**